report/clients/standalone-flow.js

```diff
@@ -0,0 +1,61 @@
+/**
+ * @license Copyright 2021 The Lighthouse Authors. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+/**
+  * @fileoverview The entry point for rendering the Lighthouse report for the HTML
+  * file created by ReportGenerator.
+  * The renderer code is bundled and injected into the report HTML along with the JSON report.
+  */
+
+import {render} from 'preact';
+import {html} from 'htm/preact';
```

Comment:
Usage of `htm` here begs the question: should we just start using `.tsx` for the report (which would allow us handle these rendering templates at compile time vs runtime)? We already have a build pipeline here, so the previous argument of "staying close to the web runtime" is no longer relevant. I am not suggesting `.jsx` because we'd still need to use `tsc` to compile the JSX, and at that point we might as well go full TS :)

Reply:
Full `.tsx` would be my preference. We are missing out on component property type checking with `htm`.

Reply:
I'm all for `.tsx` if the team is up for it :)

Reply:
ha, well, probably unsurprisingly I'll take at least a somewhat different position.

An off the shelf templating framework could be nice instead of continuously (needlessly?) rewriting our own, but DevTools and PSI are two platforms we're actively working on integrations with (with devtools as our main FR flow report platform for the time being), but this fits neither of their component or build approaches.

If we went the htm/preact (...and `tsx`) way, it would mean giving up on future easier integration with DevTools, though we could continue to be a bundled FFI there. But it would also mean any explorations of FR in PSI would be blocked on the report side.

There's also the issue of report and build complexity when we'd have two different templating frameworks at play. I think there is an argument for layering (the part that needs to be compatible with all platforms is compatible; outside of that we can be more flexible), but it's less of a strong argument because the refactor of the existing report generation is in progress _right now_ (we literally have a PR open for [the templating library we just decided to write](https://github.com/GoogleChrome/lighthouse/pull/12803) :), so the additional complexity, build time, maintenance, etc tradeoffs should be evaluated alongside the same choices we're making for the in-progress report factor, and we should be eliminating extra steps where we can.

Ideally the choice would be all or nothing. If we can't use the same stack for very similar purposes in code in the same bundle, we'd instead pick a stack that would work for that.

Reply:
> We are missing out on component property type checking with htm

is this definite? @developit has been a fan of working with tsc-checked jsdoc at least somewhat recently, and AFAIK template literal type checking works fine in JS files (that's how our [typed `querySelector`](https://github.com/g-plane/typed-query-selector/blob/360bc09d0d73352cb8f4c7a84ec07ab8a9e87d04/parser.d.ts) and `yargs` [kebab-casifying](https://github.com/GoogleChrome/lighthouse/blob/bb9871465549225fc477697b586f032475572671/types/externs.d.ts#L100-L116) work)

Reply:
> Ideally the choice would be all or nothing. If we can't use the same stack for very similar purposes in code in the same bundle, we'd instead pick a stack that would work for that.

This was not how our numerous discussions in this past on this topic have concluded. The individual report embeddable everywhere has different needs from the flow report where layering fits nicely.

> this fits neither of their component or build approaches.

Correct, and that's part of what the refactor of the report API is aiming to move toward, moving more toward a black box and severing the expectation that *every* part of the renderer is customizable. 

> I think there is an argument for layering (the part that needs to be compatible with all platforms is compatible; outside of that we can be more flexible)

This is indeed the goal.

> but it's less of a strong argument because the refactor of the existing report generation is in progress right now

Refactoring to make our universal embedded environment easier to work in is very much inline with making our non-universal app-like environment easier to work in. I'm not sure I see how the limitations of the universal environment should dictate the supporting libraries we're able to use in the standalone app-like environment. The choice of `preact/htm` was indeed to keep the difference in build process minimal, but as you note, a split is a split, and if the efficiencies gained from incremental tooling adoption outweigh the potential confusion to those who will be working on this regularly, I'm all for it.

Reply:
> is this definite? @developit has been a fan of working with tsc-checked jsdoc at least somewhat recently, and AFAIK template literal type checking works fine from JS files (that's how our typed querySelector and yargs kebab-casifying work)

`htm` uses `any` for template literal values: https://github.com/developit/htm/blob/aefd98dd9035a97f6d3bfeefcc6fc28b6eaba53d/src/integrations/preact/index.d.ts#L4.

Reply:
> > Ideally the choice would be all or nothing. If we can't use the same stack for very similar purposes in code in the same bundle, we'd instead pick a stack that would work for that.
> 
> This was not how our numerous discussions in this past on this topic have concluded. The individual report embeddable everywhere has different needs from the flow report where layering fits nicely.

Encapsulation of core rendering functions in #12772 and layering on top absolutely makes sense, but a common cause of divergence/tech debt of not updating an existing working system as a new stack is used on top is not necessarily at play here since we're updating the existing working system at the same time. I think it's worth a deliberate decision to introduce two different frameworks rather than treating it as two separate decisions.

I think one issue is I believe the last time we all discussed this was before #12803 was even an idea and we ended on the question if preact wasn't going to work for PSI, what kind of templating/state management approach would...and then we didn't answer that.